File: to-do-app/to-do-kotlin/src/main/kotlin/com/renan/todo/dto/DTO.kt
package com.renan.todo.dto

import com.renan.todo.util.STATUS_INACTIVE
import com.renan.todo.util.STATUS_INCOMPLETE

/**
 * Transfer Object for Error Message
 */
class ErrorDTO(val errorMessage: String = "")

/**
 * User data to return as login operation
 */
class UserDTO {
    var email: String = ""
    var jwt: String = ""
    var id: Int = 0
    var status: Int = STATUS_INACTIVE
    var expiresIn: Int = 0
    var name: String = ""
    var lastName: String? = null
}

/**
 * Transfer Object for Task
 */
class TaskDTO {
    var id: Int = 0
    var taskName: String = ""
    var taskDescription: String? = null
    var deadline: String? = null
    var taskStatus: Int = STATUS_INCOMPLETE
}

/**
 * Transfer Object for TaskList
 */
class TaskListDTO {
    var id: Int = 0
    var listName: String = ""
    var listDescription: String? = null
    var tasks: List<TaskDTO> = ArrayList()
}

File: to-do-app/to-do-node/src/business/dto/dto.ts
import { Task } from "../../model/task";
import { TaskList } from "../../model/taskList";
import { User } from "../../model/user";
import token from "../token/token"

export class UserDTO {
    constructor(
        public jwt: string,
        public expiresIn: number,
        public id: number,
        public name: string,
        public lastName: string | null,
        public email: string,
        public userStatus: number,
    ) { }
}

export class TaskDTO {
    constructor(
        public id: number,
        public taskName: string,
        public taskDescription: string | null,
        public deadline: string | null,
        public taskStatus: number,
    ) { }
}

export class TaskListDTO {
    constructor(
        public id: number,
        public listName: string,
        public listDescription: string | null,
        public tasks: TaskDTO[],
    ) { }
}

export const convertUserToDTO = (user: User) => {
    return new UserDTO(
        token.generateToken(user),
        token.EXP_TIME,
        user.id!,
        user.firstName,
        user.lastName,
        user.email,
        user.userStatus
    );
}

export const refreshUserDTO = (userDTO: UserDTO) => {
    return new UserDTO(
        token.refreshToken(userDTO.jwt),
        token.EXP_TIME,
        userDTO.id,
        userDTO.name,
        userDTO.lastName,
        userDTO.email,
        userDTO.userStatus
    );
}

export const convertTaskListToDTO = (list: TaskList): TaskListDTO => {
    return new TaskListDTO(
        list.id!,
        list.listName!,
        list.listDescription,
        []
    );
}

export const convertTaskListsToDTOs = (lists: TaskList[]): TaskListDTO[] => {
    let result: TaskListDTO[] = []
    lists.forEach(list => {
        result.push(convertTaskListToDTO(list))
    });
    return result;
}

export const convertTaskToDTO = (task: Task): TaskDTO => {
    return new TaskDTO(
        task.id!,
        task.taskName,
        task.taskDescription,
        task.deadline,
        task.taskStatus
    )
}

export const convertTasksToDTOs = (tasks: Task[]): TaskDTO[] => {
    let result: TaskDTO[] = []
    tasks.forEach(task => {
        result.push(convertTaskToDTO(task))
    });
    return result;
}

export const convertTasksAndListsToDTO = (lists: TaskList[], tasks: Task[]): TaskListDTO[] => {
    let tasklistsDTO: TaskListDTO[] = convertTaskListsToDTOs(lists);
    tasks.forEach(task => {
        const taskDTO = convertTaskToDTO(task);
        for (let i = 0; i < tasklistsDTO.length; i++) {
            if (tasklistsDTO[i].id === task.listID) {
                tasklistsDTO[i].tasks.push(taskDTO);
                break;
            }
        }
    });
    return tasklistsDTO;
}
